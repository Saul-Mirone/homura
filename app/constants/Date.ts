export const format = 'LLL dd, yyyy';
