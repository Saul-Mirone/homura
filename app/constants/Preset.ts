export enum Preset {
  All = 'All Items',
  Archive = 'Archive',
  Starred = 'Starred',
  Unread = 'Unread',
}
