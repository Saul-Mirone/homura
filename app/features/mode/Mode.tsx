import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Toolkit } from '../../components/SubSideBar/ToolKit';
import { AppDispatch } from '../../store';
import { selectMode, switchMode } from './modeSlice';

export const Mode: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const mode = useSelector(selectMode);

  return <Toolkit mode={mode} onSwitchMode={(m) => dispatch(switchMode(m))} />;
};
