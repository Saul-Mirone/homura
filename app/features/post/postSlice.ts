import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import type { AppThunk, RootState } from '../../store';

type State = {
  content: string;
};

const initialState: State = {
  content: '',
};

const postSlice = createSlice({
  name: 'post',
  initialState,
  reducers: {
    setContent: (state, action: PayloadAction<string>) => {
      state.content = action.payload;
    },
  },
});

export const { setContent } = postSlice.actions;

export const getPostContentById = (postId: number): AppThunk => async (
  dispatch
) => {
  const post = await channel.getPostById(postId);
  dispatch(setContent(post.content));
};

export const postReducer = postSlice.reducer;

export const selectPost = (state: RootState) => {
  const { activeId, posts } = state.list;
  const { content } = state.post;

  if (!activeId)
    return {
      activeId,
    };

  const target = posts.find((x) => x.id === activeId);

  if (!target)
    return {
      activeId,
    };

  const { title, sourceName, unread, starred, date } = target;

  const post = {
    title,
    sourceName,
    unread,
    date,
    starred,
    content,
  };

  return {
    activeId,
    post,
  };
};
