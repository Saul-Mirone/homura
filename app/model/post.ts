import {
  Association,
  BelongsToCreateAssociationMixin,
  BelongsToGetAssociationMixin,
  BelongsToSetAssociationMixin,
  DataTypes,
  Model,
  Optional,
  Sequelize,
} from 'sequelize';
import type { Source } from './source';

export interface PostAttributes {
  id: number;
  sourceId: number;
  guid: string;
  title: string;
  content: string;
  unread: boolean;
  starred: boolean;
  date: Date;
}

export type CreatePostAttributes = Optional<
  PostAttributes,
  'id' | 'starred' | 'unread'
>;

export class Post
  extends Model<PostAttributes, CreatePostAttributes>
  implements PostAttributes {
  public id!: number;

  public sourceId!: number;

  public guid!: string;

  public title!: string;

  public content!: string;

  public unread!: boolean;

  public starred!: boolean;

  public date!: Date;

  public toJSON!: () => PostAttributes;

  public getSource!: BelongsToGetAssociationMixin<Source>;

  public setSource!: BelongsToSetAssociationMixin<Source, number>;

  public createSource!: BelongsToCreateAssociationMixin<Source>;

  public static associations: {
    source: Association<Post, Source>;
  };
}

export const initPost = (sequelize: Sequelize): void => {
  Post.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      sourceId: {
        type: DataTypes.INTEGER,
        allowNull: false,
      },
      guid: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      title: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      content: {
        type: DataTypes.TEXT,
        allowNull: false,
      },
      unread: {
        type: DataTypes.BOOLEAN,
        defaultValue: true,
      },
      starred: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
      },
      date: {
        type: DataTypes.DATE,
        allowNull: false,
      },
    },
    {
      sequelize,
      tableName: 'posts',
    }
  );
};
