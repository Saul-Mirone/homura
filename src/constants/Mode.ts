export enum Mode {
    All = 'All',
    Unread = 'Unread',
    Starred = 'Starred',
}
