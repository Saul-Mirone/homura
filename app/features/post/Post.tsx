import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Reader } from '../../components/Reader';
import { Toolkit } from '../../components/Reader/Toolkit';
import { AppDispatch } from '../../store';
import { markActiveStarredAs, markActiveUnreadAs } from '../list/listSlice';
import { getPostContentById, selectPost } from './postSlice';

export const Post: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { post, activeId } = useSelector(selectPost);

  React.useEffect(() => {
    if (!activeId) return;
    dispatch(getPostContentById(activeId));
  }, [activeId, dispatch]);

  const toolkit = post ? (
    <Toolkit
      starred={post.starred}
      onSwitchStarred={(x) => dispatch(markActiveStarredAs(!x))}
      unread={post.unread}
      onSwitchUnread={(x) => dispatch(markActiveUnreadAs(!x))}
    />
  ) : null;

  return <Reader post={post} toolkit={toolkit} />;
};
