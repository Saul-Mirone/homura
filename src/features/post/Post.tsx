import React from 'react';
import { useSelector } from 'react-redux';
import { Reader } from '../../components/Reader';
import { Toolkit } from '../../components/Reader/Toolkit';
import { useActions } from '../../hooks';
import { markActiveStarredAs, markActiveUnreadAs } from '../list/listSlice';
import { getPostContentById, selectPost } from './postSlice';

export const Post: React.FC = () => {
    const { post, activeId } = useSelector(selectPost);
    const [getPostContentByIdDispatch, markActiveStarredAsDispatch, markActiveUnreadAsDispatch] = useActions([
        getPostContentById,
        markActiveStarredAs,
        markActiveUnreadAs,
    ]);

    React.useEffect(() => {
        if (!activeId) return;
        getPostContentByIdDispatch(activeId);
    }, [activeId, getPostContentByIdDispatch]);

    const onSwitchStarred = React.useCallback((x: boolean) => markActiveStarredAsDispatch(!x), [
        markActiveStarredAsDispatch,
    ]);
    const onSwitchUnread = React.useCallback((x: boolean) => markActiveUnreadAsDispatch(!x), [
        markActiveUnreadAsDispatch,
    ]);
    const onShare = React.useCallback(() => window.open(post?.link), [post?.link]);

    const toolkit = React.useMemo(
        () =>
            post ? (
                <Toolkit
                    starred={post.starred}
                    unread={post.unread}
                    onSwitchStarred={onSwitchStarred}
                    onSwitchUnread={onSwitchUnread}
                    onShare={onShare}
                />
            ) : null,
        [onShare, onSwitchStarred, onSwitchUnread, post],
    );

    return <Reader post={post} toolkit={toolkit} />;
};
