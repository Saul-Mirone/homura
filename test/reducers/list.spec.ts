/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { AnyAction } from 'redux';

// organize-imports-ignore
import { mockChannel } from '../test-tools/mockChannel';

import {
    getListByPreset,
    getListBySourceId,
    listReducer,
    markAsRead,
    markAsStarred,
    markAsUnread,
    reset,
    setActiveId,
    setFilter,
} from '../../src/features/list/listSlice';
import { post } from '../fixture/post';
import { mockStore } from '../test-tools/mockStore';
import { formatActions } from '../test-tools/formatActions';
import { Mode } from '../../src/constants/Mode';
import { Preset } from '../../src/constants/Preset';

test('should handle initial state', () => {
    expect(listReducer(undefined, {} as AnyAction)).toMatchSnapshot();
});

test('should handle reset', () => {
    expect(
        listReducer(
            {
                posts: post,
                activeId: 1,
                filter: 'filter',
            },
            {
                type: reset,
            },
        ),
    ).toMatchSnapshot();
});

test('should handle setActiveId', () => {
    expect(
        listReducer(undefined, {
            type: setActiveId,
            payload: 1,
        }),
    ).toMatchSnapshot();
});

test('should handle setFilter', () => {
    expect(
        listReducer(undefined, {
            type: setFilter,
            payload: 'filter',
        }),
    ).toMatchSnapshot();
});

describe('getListBySourceId', () => {
    it('should action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.getSourceById.mockResolvedValue(post);
        await store.dispatch(getListBySourceId({ id: 5, mode: Mode.All }));

        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should handle fulfilled', () => {
        expect(
            listReducer(undefined, {
                type: getListBySourceId.fulfilled,
                payload: post,
            }),
        ).toMatchSnapshot();
    });
});

describe('getListByPreset', () => {
    it('should action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.getPostByPreset.mockResolvedValue(post);
        await store.dispatch(getListByPreset(Preset.Unread));

        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should handle fulfilled', () => {
        expect(
            listReducer(undefined, {
                type: getListByPreset.fulfilled,
                payload: post,
            }),
        ).toMatchSnapshot();
    });
});

describe('markAsUnread', () => {
    it('should action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.setPostUnread.mockResolvedValue(undefined!);
        await store.dispatch(markAsUnread({ id: 1, unread: true }));

        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should handle fulfilled', () => {
        expect(
            listReducer(
                {
                    posts: post,
                    filter: '',
                    activeId: 1,
                },
                {
                    type: markAsUnread.fulfilled,
                    payload: {
                        id: 1,
                        unread: true,
                    },
                },
            ),
        ).toMatchSnapshot();
    });
});

describe('markAsStarred', () => {
    it('should action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.setPostStarred.mockResolvedValue(undefined!);
        await store.dispatch(markAsStarred({ id: 1, starred: false }));

        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should handle fulfilled', () => {
        expect(
            listReducer(
                {
                    posts: post,
                    filter: '',
                    activeId: 1,
                },
                {
                    type: markAsStarred.fulfilled,
                    payload: {
                        id: 1,
                        starred: false,
                    },
                },
            ),
        ).toMatchSnapshot();
    });
});

describe('markAsRead', () => {
    it('should number payload action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.markAllAsReadBySourceId.mockResolvedValue(undefined!);
        await store.dispatch(markAsRead(1));

        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should array payload action list match snapshot', async () => {
        const store = mockStore();
        mockChannel.markAllAsReadBySourceId.mockResolvedValue(undefined!);
        await store.dispatch(markAsRead([1, 2]));

        expect(formatActions(store)).toMatchSnapshot();
    });

    it('should handle fulfilled', () => {
        expect(
            listReducer(
                {
                    posts: post,
                    filter: '',
                    activeId: 1,
                },
                {
                    type: markAsRead.fulfilled,
                },
            ),
        ).toMatchSnapshot();
    });
});
