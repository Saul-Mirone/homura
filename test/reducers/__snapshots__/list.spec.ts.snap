// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getListByPreset should action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": "Unread",
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "list/getListByPreset/pending",
  },
  Object {
    "meta": Object {
      "arg": "Unread",
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": Array [
      Object {
        "date": "test-date-1",
        "icon": "test-icon-1",
        "id": 1,
        "link": "test-link-1",
        "name": "test-source-name-1",
        "sourceId": 0,
        "starred": true,
        "title": "test-title-1",
        "unread": false,
      },
      Object {
        "date": "test-date-2",
        "icon": "test-icon-2",
        "id": 2,
        "link": "test-link-2",
        "name": "test-source-name-2",
        "sourceId": 1,
        "starred": false,
        "title": "test-title-2",
        "unread": true,
      },
    ],
    "type": "list/getListByPreset/fulfilled",
  },
]
`;

exports[`getListByPreset should handle fulfilled 1`] = `
Object {
  "activeId": undefined,
  "filter": "",
  "posts": Array [
    Object {
      "date": "test-date-1",
      "icon": "test-icon-1",
      "id": 1,
      "link": "test-link-1",
      "name": "test-source-name-1",
      "sourceId": 0,
      "starred": true,
      "title": "test-title-1",
      "unread": false,
    },
    Object {
      "date": "test-date-2",
      "icon": "test-icon-2",
      "id": 2,
      "link": "test-link-2",
      "name": "test-source-name-2",
      "sourceId": 1,
      "starred": false,
      "title": "test-title-2",
      "unread": true,
    },
  ],
}
`;

exports[`getListBySourceId should action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": Object {
        "id": 5,
        "mode": "All",
      },
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "list/getListBySourceId/pending",
  },
  Object {
    "meta": Object {
      "arg": Object {
        "id": 5,
        "mode": "All",
      },
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": Array [
      Object {
        "date": "test-date-1",
        "icon": "test-icon-1",
        "id": 1,
        "link": "test-link-1",
        "name": "test-source-name-1",
        "sourceId": 0,
        "starred": true,
        "title": "test-title-1",
        "unread": false,
      },
      Object {
        "date": "test-date-2",
        "icon": "test-icon-2",
        "id": 2,
        "link": "test-link-2",
        "name": "test-source-name-2",
        "sourceId": 1,
        "starred": false,
        "title": "test-title-2",
        "unread": true,
      },
    ],
    "type": "list/getListBySourceId/fulfilled",
  },
]
`;

exports[`getListBySourceId should handle fulfilled 1`] = `
Object {
  "activeId": undefined,
  "filter": "",
  "posts": Array [
    Object {
      "date": "test-date-1",
      "icon": "test-icon-1",
      "id": 1,
      "link": "test-link-1",
      "name": "test-source-name-1",
      "sourceId": 0,
      "starred": true,
      "title": "test-title-1",
      "unread": false,
    },
    Object {
      "date": "test-date-2",
      "icon": "test-icon-2",
      "id": 2,
      "link": "test-link-2",
      "name": "test-source-name-2",
      "sourceId": 1,
      "starred": false,
      "title": "test-title-2",
      "unread": true,
    },
  ],
}
`;

exports[`markAsRead should array payload action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": Array [
        1,
        2,
      ],
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "list/markAsRead/pending",
  },
  Object {
    "meta": Object {
      "arg": Array [
        1,
        2,
      ],
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": undefined,
    "type": "list/markAsRead/fulfilled",
  },
]
`;

exports[`markAsRead should handle fulfilled 1`] = `
Object {
  "activeId": 1,
  "filter": "",
  "posts": Array [
    Object {
      "date": "test-date-1",
      "icon": "test-icon-1",
      "id": 1,
      "link": "test-link-1",
      "name": "test-source-name-1",
      "sourceId": 0,
      "starred": true,
      "title": "test-title-1",
      "unread": false,
    },
    Object {
      "date": "test-date-2",
      "icon": "test-icon-2",
      "id": 2,
      "link": "test-link-2",
      "name": "test-source-name-2",
      "sourceId": 1,
      "starred": false,
      "title": "test-title-2",
      "unread": false,
    },
  ],
}
`;

exports[`markAsRead should number payload action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": 1,
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "list/markAsRead/pending",
  },
  Object {
    "meta": Object {
      "arg": 1,
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": undefined,
    "type": "list/markAsRead/fulfilled",
  },
]
`;

exports[`markAsStarred should action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": Object {
        "id": 1,
        "starred": false,
      },
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "list/markAsStarred/pending",
  },
  Object {
    "meta": Object {
      "arg": Object {
        "id": 1,
        "starred": false,
      },
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": Object {
      "id": 1,
      "starred": false,
    },
    "type": "list/markAsStarred/fulfilled",
  },
]
`;

exports[`markAsStarred should handle fulfilled 1`] = `
Object {
  "activeId": 1,
  "filter": "",
  "posts": Array [
    Object {
      "date": "test-date-1",
      "icon": "test-icon-1",
      "id": 1,
      "link": "test-link-1",
      "name": "test-source-name-1",
      "sourceId": 0,
      "starred": false,
      "title": "test-title-1",
      "unread": false,
    },
    Object {
      "date": "test-date-2",
      "icon": "test-icon-2",
      "id": 2,
      "link": "test-link-2",
      "name": "test-source-name-2",
      "sourceId": 1,
      "starred": false,
      "title": "test-title-2",
      "unread": true,
    },
  ],
}
`;

exports[`markAsUnread should action list match snapshot 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": Object {
        "id": 1,
        "unread": true,
      },
      "requestId": "fake-request-id",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "list/markAsUnread/pending",
  },
  Object {
    "meta": Object {
      "arg": Object {
        "id": 1,
        "unread": true,
      },
      "requestId": "fake-request-id",
      "requestStatus": "fulfilled",
    },
    "payload": Object {
      "id": 1,
      "unread": true,
    },
    "type": "list/markAsUnread/fulfilled",
  },
]
`;

exports[`markAsUnread should handle fulfilled 1`] = `
Object {
  "activeId": 1,
  "filter": "",
  "posts": Array [
    Object {
      "date": "test-date-1",
      "icon": "test-icon-1",
      "id": 1,
      "link": "test-link-1",
      "name": "test-source-name-1",
      "sourceId": 0,
      "starred": true,
      "title": "test-title-1",
      "unread": true,
    },
    Object {
      "date": "test-date-2",
      "icon": "test-icon-2",
      "id": 2,
      "link": "test-link-2",
      "name": "test-source-name-2",
      "sourceId": 1,
      "starred": false,
      "title": "test-title-2",
      "unread": true,
    },
  ],
}
`;

exports[`should handle initial state 1`] = `
Object {
  "activeId": undefined,
  "filter": "",
  "posts": Array [],
}
`;

exports[`should handle reset 1`] = `
Object {
  "activeId": undefined,
  "filter": "",
  "posts": Array [],
}
`;

exports[`should handle setActiveId 1`] = `
Object {
  "activeId": 1,
  "filter": "",
  "posts": Array [],
}
`;

exports[`should handle setFilter 1`] = `
Object {
  "activeId": undefined,
  "filter": "filter",
  "posts": Array [],
}
`;
